Add backend domain changes and story node back-population job: GSoC M1.4

Kej-r03 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #16535 .
2. This PR does the following: 
 -->  It adds 5 new fields: status, planned_publication_date, last_modified, first_publication_date, unpublishing_reason, to the backend StoryNode domain model. (Many of the file changes in this PR only involve updating the StoryNodeDict objects in test files)
 -->  I have also added the beam job for back population of these 5 fields in the datastore. The logic for populating the values of these 5 fields is as follows:
   1. **status**: If the story is published, the story node will have status 'Published', else 'Draft'
   2. **unpublishing_reason**: None for all the story nodes.
   3. **first_publication_date**: If the story is not published, it is None. If story is published, we check whether the story was published later or the node was created later. Whichever happened later, we keep the date time corresponding to that event as the first_publication_date of the node.
   4. **planned_publication_date**: Same as first publication date.
   5. **last_modified**: In the version history of the StoryModel, we check the last version where any of the node's properties was updated and the corresponding date-time is the last_modified for the StoryNode. If this never happened, it will be the first_publication_date (later of either the story publication date or the node creation date)

Link to Testing Doc: https://docs.google.com/document/d/1NAWFr197gNGKDEjqBwO-u5FuZ6Nwzff226MOZGaDOis/edit?usp=sharing

Points to note:
1. In the `story_domain.py`, I have kept the fields optional for NOW, and also in the `to_dict()` and `from_dict()` functions of `StoryNode` class, I have kept them as None, to prevent errors in fetching the values of these fields in present state. After the back-population of datastore is complete, we can make the changes here from None to the actual values.
6. I have not written the test file for the job. I will be able to write it is as soon as the correctness of the job itself is confirmed.


## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

Demo video proof of job.

https://github.com/oppia/oppia/assets/99071926/10484c0d-aa2f-4a5a-869e-d38992d71f37

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Kej-r03 committed:
Add backend domain changes and story node back-population job

lkbhitesh07 reviewed (changes requested):
Hey @Kej-r03, thanks for the changes. Few notes:
1. Proof of changes is missing from the description, you have to run the job on the release-coordinator page and make sure that the changes work as expected, and post the screenshots here.
2. Whenever you are writing a job that makes changes to the datastore you have to write another job similar to it that does not make the changes to the datastore, which means you have to avoid doing this part `'Put models into datastore' >> ndb_io.PutModels()` in the job, have a look at other jobs for reference.
3. How would you verify that the changes you have done to the datastore are correct? There should be some way to verify that, I think there can be 2 ways but please discuss this with your mentors as well:
(a) Write a job that will try to fetch the newly added fields and check if everything works correctly.
(b) Ask the admin to fetch the data from the datastore and check the values for verification.
I would say go ahead with the (a), I think @DubeySandeep can guide us here.
4. Please add the tests for the job, I think in case some edits are suggested they will be regarding the logic you have wrote and it will not affect the tests that much.

Thanks

DubeySandeep commented:
> 3. How would you verify that the changes you have done to the datastore are correct? There should be some way to verify that, I think there can be 2 ways but please discuss this with your mentors as well:
(a) Write a job that will try to fetch the newly added fields and check if everything works correctly.
(b) Ask the admin to fetch the data from the datastore and check the values for verification.
I would say go ahead with the (a), I think @DubeySandeep can guide us here.

@lkbhitesh07 Thanks for raising it, I think this goes under the pre/post checks of "job request doc", you can find the details on the wiki page to run jobs on the backup server/prod. 

Also, make sure to write an audit job corresponding to any new job added to the codebase. @vojtechjelinek Do we have any general pattern for writing audit jobs without duplicate code?

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Write story node job test

Kej-r03 commented:
@lkbhitesh07 I have written the test, but it is just failing due to one reason, that when we fetch the model back to verify the values after running the job, it does not fetch the updated model. Instead, if I put the model in the job using put() function (rather than ndb_io.PutModels()), then it is fetching the updated models. What do you suggest for this?

Also regarding the screenshots, actually I can post them, but one cannot verify if they are correct values, rather you can carefully read my code logic, and also see its test file for example (which is giving correct results with put() as described above)

Kej-r03 commented:
@vojtechjelinek Please reply to the Sandeep's comment, so that I can proceed accordingly. Thanks.

lkbhitesh07 commented:
> @lkbhitesh07 I have written the test, but it is just failing due to one reason, that when we fetch the model back to verify the values after running the job, it does not fetch the updated model. Instead, if I put the model in the job using put() function (rather than ndb_io.PutModels()), then it is fetching the updated models. What do you suggest for this?

Hmm, have you tried running the job on the release coordinator page and then check the state of the models using [dsadmin](https://github.com/oppia/oppia/wiki/Debugging-datastore-locally)?
I'm not sure why the `ndb_io.PutModels()` is not working in your case because we've been using this only to update the datastore. You have to debug things, looking at other jobs might help. Maybe try one thing to debug at what stage things are:
a. Visit the develop branch and create the topic, story etc models.
b. Visit to your branch where you have changes and then run the job from the release-coordinator page.
c. Visit the dsadmin to check the state of the model, if you are able to see the changes or not.
Also please try to look at other jobs that are using PutModels and compare what might be wrong.

> Also regarding the screenshots, actually I can post them, but one cannot verify if they are correct values, rather you can carefully read my code logic, and also see its test file for example (which is giving correct results with put() as described above)

The screenshots are needed in order to verify that the job will run successfully on the backup server and that there is no issue in running the job. To verify the logic you can also paste the screenshots of "before" and "after" of the datastore using dsadmin.

Thanks

vojtechjelinek commented:
> > 3. How would you verify that the changes you have done to the datastore are correct? There should be some way to verify that, I think there can be 2 ways but please discuss this with your mentors as well:
> >    (a) Write a job that will try to fetch the newly added fields and check if everything works correctly.
> >    (b) Ask the admin to fetch the data from the datastore and check the values for verification.
> >    I would say go ahead with the (a), I think @DubeySandeep can guide us here.
> 
> @lkbhitesh07 Thanks for raising it, I think this goes under the pre/post checks of "job request doc", you can find the details on the wiki page to run jobs on the backup server/prod.
> 
> Also, make sure to write an audit job corresponding to any new job added to the codebase. @vojtechjelinek Do we have any general pattern for writing audit jobs without duplicate code?

You can see an example of an integration of normal and audit job into one class in https://github.com/oppia/oppia/blob/develop/core/jobs/batch_jobs/skill_migration_jobs.py#L57 but it really depends on what are the things that you plan to audit, if it is just if some class can be loaded (then the aforementioned approach will work) but sometimes you want to do more complicated one-time auditing in the job and then it might make more sense to write it separately.

Nik-09 commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,162 @@
+                topic_model = topic_models.TopicModel.get(
+                    story_model.corresponding_topic_id)
+                story_reference = next(
+                    x for x in topic_model.canonical_story_references
```
Can we use a more specific variable name?

Kej-r03 replied:
Done

Nik-09 commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,162 @@
+                    x for x in topic_model.canonical_story_references
+                        if x['story_id'] == story_model.id)
+                for node in nodes:
+                    node['unpublishing_reason'] = None
```
Can we set node['status'] to 'Draft' as a default and remove the else part?

Kej-r03 replied:
Done

Nik-09 reviewed (changes requested):
Thank you @Kej-r03 for the PR, I have left 2 smaller comments, leaving the logic review for the beam jobs to code owners. Thank you

kevintab95 commented on core/utils.py:
```diff
@@ -623,6 +623,20 @@ def get_time_in_millisecs(datetime_obj: datetime.datetime) -> float:
     return msecs + (datetime_obj.microsecond / 1000.0)
 
 
+def convert_milisecs_time_to_datetime_object(
```
typo: milliseconds

Perhaps you can use msecs because I see that used often in the codebase.

kevintab95 replied:
Ditto below.

Kej-r03 replied:
Done

kevintab95 commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,162 @@
+        story_models.StoryModel,
+        Tuple[str, Exception]
+    ]:
+        """Populate the new fields in each story node of the StoryModel
+        instance.
```
It is unclear what "new" here means. Prefer being as precise in the doc string as possible e.g. what are the fields being populated.

Kej-r03 replied:
Done

kevintab95 reviewed (changes requested):
Thanks @Kej-r03! Just a couple of minor comments. Could you please confirm the following?
1. Make sure that the "proof of changes" are updated. Given that this isn't a user facing change, it should be fine to provide some screenshots of how this was tested locally (e.g. dasdmin shows new fields populated correctly, the job outputs the correct results without errors).
2. Submit a job run request once the PR is ready (see [wiki](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)).

Thanks!

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Debug beam jobs test

Kej-r03 commented:
Here's the video demo of the beam jobs test. Also now the beam job test is running successfully.

https://github.com/oppia/oppia/assets/99071926/10484c0d-aa2f-4a5a-869e-d38992d71f37

Kej-r03 commented:
@lkbhitesh07 @kevintab95 @Nik-09 I have made the changes as per your comments. PTAL.

lkbhitesh07 reviewed (changes requested):
Hey @Kej-r03, thanks for the changes. I think there is still a few things missing, can you please perform the following and assign me again:
- Thanks for the proof of work but make sure to put that in the PR description.
- Can you also make a video according to the steps I mentioned in this [comment](https://github.com/oppia/oppia/pull/18360#issuecomment-1588932455), this way we can make sure that the datastore will be fine when we run your job?
- Still, the audit job is missing which is the same as the original beam job but does not make changes to the datastore.
- How will you verify that the changes you have made are done correctly.
Thanks

**EDIT** Also please fix the backend tests, I think you have to add your job test file to `backend_test_shards.json`

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Add audit beam job

Kej-r03 commented:
Hi @lkbhitesh07 I have updated the PR description with the video proof. 

Also I have followed the same steps in the video as you have described in that comment. You can see 3 chapters in a story, and the corresponding page in dsadmin for the StoryModel, and after running the beam job, the 5 new fields are gettting updated to their proper values.

I have pushed the audit job commit now, please take a look at that.

And also as I had said earlier, it will be difficult to verify the correctness of changes from the video proof, as one cannot know when I had updated the chapter or published the story in the past. For this reason, you can go through the logic of my code once, and also read the sample example in the corresponding test, which is passing, to check the correctness of the logic.

Nik-09 reviewed (approved):
LGTM, Thanks @Kej-r03

kevintab95 reviewed (changes requested):
Thanks @Kej-r03! Leaving a review hold until this has been verified on the server.

lkbhitesh07 reviewed (changes requested):
Great work! Just a small nit. So to avoid duplication with the audit job you have added, you can do something like [this](https://github.com/oppia/oppia/blob/5957b9ddc0b58aa61477a3685195b39ac015dfe4/core/jobs/batch_jobs/translation_migration_jobs.py#L195). Other than that everything looks good.
Thanks
**EDIT:** Please fix the backend tests, these are the verifications that your code works fine in the codebase.

gp201 commented:
@Kej-r03 Please assign after the backend tests have been fixed. Thanks!

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Fix backend tests

Kej-r03 commented:
@lkbhitesh07 I have written the audit job as per the link. Also the backend test is fixed now.
@gp201 PTAL.

Kej-r03 committed:
Fix backend coverage

lkbhitesh07 reviewed (approved):
Thanks for the changes. Approving since @kevintab95 already has a review hold until it gets tested on the server.
Please fix the backend tests and get all the reviews asap, before that it cannot be tested on the backup server.
Thanks

gp201 commented:
@Kej-r03 Please fix the backend tests and assign me. Thanks!

```
Ran 1453 tests in 79 test classs.
All tests passed.

INCOMPLETE PER-FILE COVERAGE (98.0%): core.jobs.batch_jobs.story_node_jobs_test
Name                                      Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------------------------------
core/jobs/batch_jobs/story_node_jobs.py      69      0     30      2    98%   75->exit, 105->128
-------------------------------------------------------------------------------------
TOTAL                                        69      0     30      2    98%
```

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Fix backend coverage

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Feature gate backend domain functions

Kej-r03 committed:
Fix backend tests

Kej-r03 committed:
Fix backend tests

Kej-r03 committed:
Fix story domain coverage

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 commented:
@gp201 @kevintab95 Now all the backend tests are fixed. Please review and test this on the backup server.
(Please do this as early as possible because it is a dependency PR, the next PRs cannot be opened until this is merged. Thanks)

gp201 reviewed (approved):
LGTM for codeowner files

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Fix backend shard error

Kej-r03 commented:
@kevintab95 A reminder to review this PR. This PR is blocking my upcoming GSoC PRs so please help me to get this merged. 
The failing test is probably a random flake, but all the backend tests are passing. Thanks

kevintab95 commented:
Hi @Kej-r03, have you submitted a request per https://github.com/oppia/oppia/pull/18360/#issuecomment-1585740626 ?

This is required to run the job on the server.

Kej-r03 commented:
Oh i did not know about that. I filled that form just now.

DubeySandeep commented on core/domain/story_domain.py:
```diff
@@ -549,7 +650,12 @@ def to_dict(self) -> StoryNodeDict:
             'outline': self.outline,
             'outline_is_finalized': self.outline_is_finalized,
-            'exploration_id': self.exploration_id
+            'exploration_id': self.exploration_id,
+            'status': None,
+            'planned_publication_date_msecs': None,
+            'last_modified_msecs': None,
+            'first_publication_date_msecs': None,
+            'unpublishing_reason': None
```
WHya re we setting these to null? Are we expecting to be None even after running the population job?

DubeySandeep replied:
I see the following in the PR description:
> I have kept them as None, to prevent errors in fetching the values of these fields in present state. 

(I'm assuming these are fields of the model and not part of jsonProperty) I think it will be set to None by default if it doesn't exist. Also, `self` will definitely have these fields (through constructor) so I'm not sure why are we setting it to None.

Kej-r03 replied:
Actually they are None because their values currently don't exist in datastore models.
I mentioned them explicitly as None here, so that after the population is done, I can simply the change the values from None to  the corresponding fields like self.status, self.unpublishing_reason, etc.
Also, `self` will not be having the values of these fields because they are not yet present in the existing models.

DubeySandeep replied:
> Also, self will not be having the values of these fields because they are not yet present in the existing models.

Yeah, so in case these properties exist in self irrespective of whether it's null or other value, use self.unpublishing_reason. This way you won't have to initiate another PR to change the to_dict method.

Note: You'll have to add validation checks for these new fields and validate it before populating the fields in models.

Kej-r03 replied:
Done

DubeySandeep commented on core/domain/story_domain.py:
```diff
@@ -1805,6 +1913,72 @@ def update_node_prerequisite_skill_ids(
         self.story_contents.nodes[node_index].prerequisite_skill_ids = (
             new_prerequisite_skill_ids)
```
ARe we expecting to add validation for the new fields?

Kej-r03 replied:
I am not sure if there is any validation required as such. Do you have any suggestions?

DubeySandeep replied:
Yeah, you'll have to add validation for new fields please check other domain object classes for reference.

Kej-r03 replied:
Done

DubeySandeep commented on core/domain/story_domain.py:
```diff
@@ -1805,6 +1913,72 @@ def update_node_prerequisite_skill_ids(
+
+        Args:
+            node_id: str. The id of the node.
+            new_status: str. The new publication status of the node.
```
```suggestion
        """Updates the status of a given node.

        Args:
            node_id: str. The Id of the node.
            new_status: str. The new publication status of the node.
```

DubeySandeep replied:
Follow this elsewhere*

Kej-r03 replied:
Updated.

DubeySandeep commented on core/domain/story_domain.py:
```diff
@@ -1805,6 +1913,72 @@ def update_node_prerequisite_skill_ids(
+        Args:
+            node_id: str. The id of the node.
+            new_planned_publication_date_msecs: float. The planned publication
+                date for the node in miliseconds.
```
_of_ the node*

Kej-r03 replied:
Updated.

DubeySandeep commented on core/domain/story_domain.py:
```diff
@@ -1805,6 +1913,72 @@ def update_node_prerequisite_skill_ids(
+        Args:
+            node_id: str. The id of the node.
+            new_last_modified_msecs: float. The last modified date time
+                for the node in miliseconds.
```
_of_ the node

Kej-r03 replied:
Updated.

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+
+
+class PopulateStoryNodeJob(base_jobs.JobBase):
+    """Job that migrates story nodes, for serial chapter launch feature"""
```
Should we add more description i.e, which field it populates?

Kej-r03 replied:
Updated.

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+    DATASTORE_UPDATES_ALLOWED = True
+
+    def _update_story_node(
+            self, story_model: story_models.StoryModel
```
are we expecting 4 spaces for indentation here and not 8?

Kej-r03 replied:
Changed to 4 space indentation.

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+
+        try:
+            with datastore_services.get_ndb_context():
+                topic_model = topic_models.TopicModel.get(
```
Why do we need to make get request here? I'm assuming the total number of request would be O(m*n) here [m=nuber of stories in topic, n=number of topics]

Can we make parallel request in pipe to fetch all the topics at once and create a tuple of (topic, [stories]) and traverse through each item of the tuple?

Kej-r03 replied:
Yes that can be done. Actually I am writing beam jobs for the first time, so I wrote this in a simplistic manner. I am not sure of how we can create a tuple of topic and stories array using the PCollections for topic models and story models. It will be helpful if you can share some clues about it.

DubeySandeep replied:
@Kej-r03 Please check other *_jobs.py file for reference (e.g. story_migration_jobs.py). Also, go through the wiki page you'll find important information. Writing the first job feels daunting but it's not that difficult, I think it's just that the syntax/pattern is a little different than general Python code which makes it look difficult (not sure!). Let me know if you have any specific questions, I'll be more than happy to help! :)

DubeySandeep replied:
Also, you can read the official ref doc which contains good examples:
https://beam.apache.org/documentation/transforms/python/
https://beam.apache.org/documentation/transforms/python/elementwise/

Kej-r03 replied:
Done

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+                        snapshot_id = topic_model.get_snapshot_id(
+                            topic_model.id, version)
+                        topic_metadata = (
+                            topic_models.TopicSnapshotMetadataModel.get(
```
Is it possible to avoid making so many get requests in the datastore? (I'm assuming we are making ~n^2 requests here)

Kej-r03 replied:
Yes I understand. I actually had given much thought to optimize this and also discussed this with Hitesh, but this was the best we could decide. Because for each node of the story, we need to traverse back the topic version history as well as story version history to get the relevant information for the node.
I thought that we had 2 options: 
1. Either to do like this and continue fetching the previous version models till the requirement is met and break the loop
2. Or to fetch all the versions of the topic and that story at the start, and then loop till we get the relevant data and break the loop. However doing this way fetches even those earlier versions which might never be needed at all for any of the nodes, and also maintaining the PCollection array of topic version models and story version models for each topic and story also appears to be a complex data structure, which looks difficult to work with.

So for simplicity and clear understandablity of the code, I have written like this.

If you have any alternative ideas to implement this, please share a description so that we can work on it.

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+                        snapshot_id = story_model.get_snapshot_id(
+                                story_model.id, version)
+                        story_metadata = (
+                            story_models.StorySnapshotMetadataModel.get(
```
Ditto*

Kej-r03 replied:
I have replied to that comment above. PTAL

DubeySandeep commented on core/utils.py:
```diff
@@ -623,6 +623,20 @@ def get_time_in_millisecs(datetime_obj: datetime.datetime) -> float:
 
+def convert_millisecs_time_to_datetime_object(
+        date_time_msecs: float) -> datetime.datetime:
+    """Returns date time object from the date time in milliseconds.
```
```suggestion
    """Returns the datetime object from the given date time in milliseconds.
```

Kej-r03 replied:
Done

DubeySandeep commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,183 @@
+                story_models.StoryModel.get_all())
+            | 'Update story node fields' >> beam.Map(self._update_story_node)
+        )
+
```
(Curious) I've reviewed a similar job PR recently, whether it was your PR? (not sure!)

Kej-r03 replied:
Yes you have reviewed this PR once, but you did not point any detailed corrections back then.

DubeySandeep replied:
Can you please share the link to the PR?

Kej-r03 replied:
No, I was saying that you have reviewed this very PR in the past, not any other PR of mine.

DubeySandeep reviewed (changes requested):
@Kej-r03 I took a pass on the PR and left a few comments (mostly related to the new jobs), can you please take a look?

JayVivarekar reviewed (changes requested):
Hi @Kej-r03! I will help you to test the job on the backup server. I am adding a blocking review till this is tested. Can you assign me after you have adressed all requested changes?

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Modify function descriptions

Kej-r03 commented:
@DubeySandeep I have answered your comments and made some changes, PTAL.

Kej-r03 commented:
There is one more concern about this PR. Actually it consists of the backend domain changes parts + beam jobs parts. I think the backend changes can be approved soon probably, but the beam jobs part may take more time (since I am writing beam jobs for the first time), and this PR is blocking my upcoming 3 GSoC PRs, that depend on the backend domain parts only. 
So is there any way that I can split the two parts of the PR (or any other workaround) to get the backend domain parts merged on an urgent basis so I can open the next 3 PRs. 
Thanks.

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Add story node validation

Kej-r03 commented:
@DubeySandeep I have made the backend domain changes in this commit (the validations of the new fields).
Can I open a new PR that has only the backend domain changes of this PR, not the beam job changes, so that it can be merged separately (and probably sooner), from the beam job parts, due to the reason I mentioned above, meanwhile I am working on the beam job.

Kej-r03 committed:
Change beam job structure to reduce datastore fetches

Kej-r03 commented:
@DubeySandeep I have made all the changes you have requested. PTAL

Also assigning this to @JayVivarekar since I have addressed all requested changes.

DubeySandeep commented:
> Can I open a new PR that has only the backend domain changes of this PR, not the beam job changes, so that it can be merged separately (and probably sooner)

Depends on whether these changes can be deployed on production separately and won't create a problem for users and corrupt existing data in any way. If not, then I will definitely suggest having a separate small PR. (In general, you should divide your projects into small PRs, so that it's easy to work on them, helpful for reviewers, and can be merged quickly!)

Kej-r03 commented:
> Depends on whether these changes can be deployed on production separately and won't create a problem for users and corrupt existing data in any way. If not, then I will definitely suggest having a separate small PR. (In general, you should divide your projects into small PRs, so that it's easy to work on them, helpful for reviewers, and can be merged quickly!)

@DubeySandeep I have opened a new PR for the backend domain parts.
Please also review the beam job changes that I had pushed yesterday.

Kej-r03 committed:
Resolve merge conflicts

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

DubeySandeep commented:
@Kej-r03 No I'm not sure whether I've to review this PR or the other PR. Also, are we sure that deploying only the domain changes won't create problem for the end user or corrupt the existing data?

Kej-r03 commented:
@DubeySandeep you can review this PR only, the other PR is already assigned to other reviewers there.
Also, after deployment, the domain changes will add these 5 new fields to those story models in datastore in which any changes are made by the curriculum admins after merging. All the 5 fields will have value None, and this will not break the current working in any manner, since these are being retrieved and used under feature flags in the frontend and they also do not interfere with any backend functionality.
Once the beam job is successfully executed, all these None values of the updates stories will get replaced by their correct values, but even before running the beamjob for the time being, the software works fine.

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

JayVivarekar commented:
@Kej-r03 Can you add a link to the testing doc in the PR description?

kevintab95 reviewed (approved):
LGTM for codeowner files. @JayVivarekar will be handling the job run and will approve once it has been successfully run. Thanks!

Kej-r03 commented:
@JayVivarekar I have added the doc link. PTAL.

JayVivarekar commented:
@Kej-r03 I have a left a comment on the doc. PTAL

Kej-r03 commented:
@JayVivarekar I have addressed the doc comment. PTAL

JayVivarekar commented on core/jobs/batch_jobs/story_node_jobs.py:
```diff
@@ -0,0 +1,210 @@
+        updated_story_models_results = (
+            topic_story_pairs
+            | 'Update story node fields' >> beam.Map(self._update_story_node)
+        )
```
@Kej-r03. You are not filtering erroneous results after running `_update_story_node`. We would like to commit nothing in case any errors are encountered. There is a transform defined in `core/jobs/transforms/results_transforms.py` which does this and is used in existing jobs. PTAL and update here accordingly. Here is the story migration job reference:

https://github.com/oppia/oppia/blob/6d78839b605f3b009b93f1a41ecb66292ddb3b6b/core/jobs/batch_jobs/story_migration_jobs.py#L188

Kej-r03 replied:
Done

JayVivarekar reviewed (changes requested):
@Kej-r03 PTAL. I noticed this while reviewing your job run doc.

DubeySandeep reviewed (approved):
Unblocking PR as @JayVivarekar is testing this PR on the backup server.

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 committed:
drain results on error fix

Kej-r03 commented:
Hi @JayVivarekar I have been really late in replying back due college exams and other simultaneous GSoC PRs. I have made the changes now. PTAL

JayVivarekar reviewed (commented):
Thanks @Kej-r03! Changes look good. Will test this ASAP.

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 commented:
Hi @JayVivarekar any updates on the beam job run?

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

JayVivarekar commented:
@Kej-r03 Backend tests are failing. PTAL

Kej-r03 committed:
Add job name in valid imports

Kej-r03 committed:
Merge branch 'develop' into story-node-backend-domain

Kej-r03 commented:
@JayVivarekar any updates on the job testing?

JayVivarekar reviewed (approved):
LGTM! Tested successfully on the backup server.
